lmorpho/lsystem.cpp

```diff
@@ -0,0 +1,306 @@
+            break;
+        case param_type::normal:
+            normal_dist = std::normal_distribution<result_type>(p_.mu, p_.sigma);
+            break;
+        default: ;
+        }
+    }
+};
+
+class burke_correlated_r {
+    double a_;
+    lsys_distribution r_;
+
+public:
+    struct result_type { double rho1, rho2; };
+
+    burke_correlated_r(double a, lsys_distribution_param r): a_(a), r_(r) {}
+
+    template <typename Gen>
+    result_type operator()(Gen& g) const {
+        double r1 = r_(g);
+        double r2 = r_(g);
+        return { r1+a_*r2, r2+a_*r1 };
+    }
+};
+
+class burke_exp_test {
+    double k1_;
+    double k2_;
+
+public:
+    using result_type = int;
+
+    burke_exp_test(double k1, double k2): k1_(k1), k2_(k2) {}
+
+    template <typename Gen>
+    result_type operator()(double delta_l, double radius, Gen& g) const {
+        std::bernoulli_distribution B(delta_l*k1_*std::exp(k2_*radius*2.0));
+        return B(g);
+    }
+};
+
+class burke_exp2_test {
+    double ak1_;
+    double ak2_;
+    double bk1_;
+    double bk2_;
+
+public:
+    using result_type = int;
+
+    burke_exp2_test(double ak1, double ak2, double bk1, double bk2):
+        ak1_(ak1), ak2_(ak2), bk1_(bk1), bk2_(bk2) {}
+
+    template <typename Gen>
+    result_type operator()(double delta_l, double radius, Gen& g) const {
+        double diam = 2.*radius;
+        std::bernoulli_distribution B(
+            delta_l*std::min(ak1_*std::exp(ak2_*diam), bk1_*std::exp(bk2_*diam)));
+
+        return B(g);
+    }
+};
+
+// L-system parameter set with instantiated distributions.
+struct lsys_sampler {
+    // Create distributions once from supplied parameter set.
+    explicit lsys_sampler(const lsys_param& P):
+        diam_soma(P.diam_soma),
+        n_tree(P.n_tree),
+        diam_initial(P.diam_initial),
+        length_step(P.length_step),
+        roll_initial(P.roll_initial),
+        pitch_initial(P.pitch_initial),
+        roll_section(P.roll_section),
+        pitch_section(P.pitch_section),
+        taper(P.taper),
+        roll_at_branch(P.roll_at_branch),
+        branch_angle(P.branch_angle),
+        diam_child(P.diam_child_a, P.diam_child_r),
+        term_test(P.pterm_k1, P.pterm_k2),
+        branch_test(P.pbranch_ov_k1, P.pbranch_ov_k2,
+            P.pbranch_nov_k1, P.pbranch_nov_k2),
+        max_extent(P.max_extent)
+    {}
+
+    lsys_distribution diam_soma;
+    lsys_distribution n_tree;
+    lsys_distribution diam_initial;
+    lsys_distribution length_step;
+    lsys_distribution roll_initial;
+    lsys_distribution pitch_initial;
+    lsys_distribution roll_section;
+    lsys_distribution pitch_section;
+    lsys_distribution taper;
+    lsys_distribution roll_at_branch;
+    lsys_distribution branch_angle;
+    burke_correlated_r diam_child;
+    burke_exp_test term_test;
+    burke_exp2_test branch_test;
+    double max_extent;
+};
+
+struct section_tip {
+    section_point p = {0., 0., 0., 0.};
```

Comment:
It is clear when we look at the definition, but when the user of this type will see something like
```
// prototype
section_tip foo();

// ...
auto tip = foo();
bar(t.p); // what is p?
bar(t.point); // much clearer
```